<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>12&nbsp; Evaluation for unsupervised machine learning</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../materials/pca_tsne_notwork.html" rel="next">
<link href="../materials/missing_data.html" rel="prev">
<link href="../site_libs/quarto-contrib/quarto-project/cambiotraining/courseformat/img/university-of-cambridge-favicon.ico" rel="icon">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-fd4369bcce5669dc6f092a5d180a1716.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script src="../site_libs/quarto-contrib/iconify-2.1.0/iconify-icon.min.js"></script>
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../site_libs/quarto-contrib/quarto-project/cambiotraining/courseformat/img/university_crest_reversed.png" alt="" class="navbar-logo light-content">
    <img src="../site_libs/quarto-contrib/quarto-project/cambiotraining/courseformat/img/university_crest_reversed.png" alt="" class="navbar-logo dark-content">
    </a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Applied Unsupervised Machine Learning</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/cambiotraining/ml-unsupervised"> 
<span class="menu-text"><iconify-icon role="img" inline="" icon="fa6-brands:github" style="font-size: 1.25em;" aria-label="Icon github from fa6-brands Iconify.design set." title="CRIT GitHub"></iconify-icon></span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://bsky.app/profile/bioinfocambs.bsky.social"> 
<span class="menu-text"><iconify-icon role="img" inline="" icon="fa6-brands:bluesky" style="font-size: 1.25em;" aria-label="Icon bluesky from fa6-brands Iconify.design set." title="CRIT Bluesky"></iconify-icon></span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://genomic.social/@BioInfoCambs"> 
<span class="menu-text"><iconify-icon role="img" inline="" icon="fa6-brands:mastodon" style="font-size: 1.25em;" aria-label="Icon mastodon from fa6-brands Iconify.design set." title="CRIT Mastodon"></iconify-icon></span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://bioinfotraining.bio.cam.ac.uk/"> 
<span class="menu-text"><iconify-icon role="img" inline="" icon="mdi:web" style="font-size: 1.25em;" aria-label="Icon web from mdi Iconify.design set." title="CRIT Website"></iconify-icon></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <a class="flex-grow-1 no-decor" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
          <h1 class="quarto-secondary-nav-title"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Evaluation for unsupervised machine learning</span></h1>
        </a>     
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
      <a href="../index.html" class="sidebar-logo-link">
      </a>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Welcome</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Applied Unsupervised Machine Learning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Data &amp; Setup</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">References and Resources</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Session 1: Introduction to Unsupervised Learning</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/walkthrough.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Introduction to Unsupervised Learning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/python_refresher.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Refresher on Python</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/normalization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Normalizing your data and PCA</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/goal_unsupervised.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Goal of Unsupervised Learning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/conceptual_basis_PCA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Conceptual and mathematical basis of PCA</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Session 2: Dimensionality Reduction</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/tsne.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">tSNE</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Session 3: Clustering</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Clustering</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/missing_data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Missing values in unsupervised machine learning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/evaluation.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Evaluation for unsupervised machine learning</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Session 4: Applications of Unsupervised Learning</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/pca_tsne_notwork.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">When PCA or tSNE might not work</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/applications.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Hands-on exercises (Applications of unsupervised machine learning)</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true">
 <span class="menu-text">Bonus Material</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/umap.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">UMAP (Optional)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/bonus_material.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Mathematical Basics and Additional Material</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#conceptual-framing" id="toc-conceptual-framing" class="nav-link active" data-scroll-target="#conceptual-framing"><span class="header-section-number">12.1</span> Conceptual framing</a></li>
  <li><a href="#quick-metrics-cheat-sheet-what-they-tell-you" id="toc-quick-metrics-cheat-sheet-what-they-tell-you" class="nav-link" data-scroll-target="#quick-metrics-cheat-sheet-what-they-tell-you"><span class="header-section-number">12.2</span> Quick metrics cheat-sheet (what they tell you)</a></li>
  <li><a href="#optional-exercise-on-cancer-data" id="toc-optional-exercise-on-cancer-data" class="nav-link" data-scroll-target="#optional-exercise-on-cancer-data"><span class="header-section-number">12.3</span> (Optional) Exercise on cancer data</a>
  <ul class="collapse">
  <li><a href="#exercise-try-another-linkage-method-and-another-distance-metric" id="toc-exercise-try-another-linkage-method-and-another-distance-metric" class="nav-link" data-scroll-target="#exercise-try-another-linkage-method-and-another-distance-metric"><span class="header-section-number">12.3.1</span> Exercise: try another linkage method and another distance metric</a></li>
  </ul></li>
  <li><a href="#evaluating-the-quality-of-clusters" id="toc-evaluating-the-quality-of-clusters" class="nav-link" data-scroll-target="#evaluating-the-quality-of-clusters"><span class="header-section-number">12.4</span> Evaluating the Quality of Clusters</a>
  <ul class="collapse">
  <li><a href="#internal-evaluation" id="toc-internal-evaluation" class="nav-link" data-scroll-target="#internal-evaluation"><span class="header-section-number">12.4.1</span> 1. Internal Evaluation</a></li>
  <li><a href="#external-evaluation" id="toc-external-evaluation" class="nav-link" data-scroll-target="#external-evaluation"><span class="header-section-number">12.4.2</span> 2. External Evaluation</a></li>
  <li><a href="#visual-evaluation" id="toc-visual-evaluation" class="nav-link" data-scroll-target="#visual-evaluation"><span class="header-section-number">12.4.3</span> 3. Visual Evaluation</a></li>
  <li><a href="#internal-evaluation-silhouette-score" id="toc-internal-evaluation-silhouette-score" class="nav-link" data-scroll-target="#internal-evaluation-silhouette-score"><span class="header-section-number">12.4.4</span> Internal Evaluation: Silhouette Score</a></li>
  <li><a href="#external-evaluation-adjusted-rand-index-ari" id="toc-external-evaluation-adjusted-rand-index-ari" class="nav-link" data-scroll-target="#external-evaluation-adjusted-rand-index-ari"><span class="header-section-number">12.4.5</span> External Evaluation: Adjusted Rand Index (ARI)</a></li>
  <li><a href="#intuition-of-silhouette-score" id="toc-intuition-of-silhouette-score" class="nav-link" data-scroll-target="#intuition-of-silhouette-score"><span class="header-section-number">12.4.6</span> Intuition of silhouette score</a></li>
  </ul></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary"><span class="header-section-number">12.5</span> Summary</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Evaluation for unsupervised machine learning</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Tip</span>Learning Objectives
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>How to evaluate unsupervised machine learning techniques</li>
<li>Know the difference between <strong>internal</strong>, <strong>external</strong>, and <strong>biological</strong> evaluation.</li>
<li>Be able to compute and interpret common metrics (silhouette).</li>
<li>Be able to choose metrics depending on whether you have ground truth, partial labels, or purely exploratory goals.</li>
<li>Communicate results in biologically meaningful ways (marker genes).</li>
</ul>
</div>
</div>
<section id="conceptual-framing" class="level2" data-number="12.1">
<h2 data-number="12.1" class="anchored" data-anchor-id="conceptual-framing"><span class="header-section-number">12.1</span> Conceptual framing</h2>
<ul>
<li><strong>Internal metrics</strong>: use only the data + clustering labels. Measure compactness vs separation (e.g., silhouette).</li>
<li><strong>External metrics</strong>: require ground truth/labels (experimental groups, annotated cell types). <!--Use ARI, NMI, precision/recall on pairwise same/different labels.--></li>
<li><strong>Biological validation</strong>: compare clusters to known marker genes, pathways, experimental metadata (batch, donor), or enrichment tests. Often the most important for biologists.</li>
</ul>
<hr>
</section>
<section id="quick-metrics-cheat-sheet-what-they-tell-you" class="level2" data-number="12.2">
<h2 data-number="12.2" class="anchored" data-anchor-id="quick-metrics-cheat-sheet-what-they-tell-you"><span class="header-section-number">12.2</span> Quick metrics cheat-sheet (what they tell you)</h2>
<ul>
<li><strong>Explained variance (PCA)</strong> — fraction of variance captured by components (useful for dimensionality reduction decisions).</li>
<li><strong>Silhouette score</strong> (−1..1) — how well each sample fits its cluster vs nearest other cluster; good general-purpose internal metric. <!--
* **Calinski–Harabasz** — ratio of between/within dispersion (higher = better).
* **Davies–Bouldin** (lower = better) — average similarity between each cluster and its most similar one.
* **Adjusted Rand Index (ARI)** — similarity between two labelings corrected for chance (commonly 0..1).
* **Normalized Mutual Information (NMI)** — information overlap between labelings (0..1).
* **Trustworthiness** (for embeddings like UMAP/t-SNE) — how well local neighborhoods are preserved.
--></li>
<li><strong>Stability / reproducibility</strong> — how consistent cluster assignments are under parameter changes.</li>
</ul>
<hr>
</section>
<section id="optional-exercise-on-cancer-data" class="level2" data-number="12.3">
<h2 data-number="12.3" class="anchored" data-anchor-id="optional-exercise-on-cancer-data"><span class="header-section-number">12.3</span> (Optional) Exercise on cancer data</h2>
<ul>
<li><p>You have been given some data on cancer cell lines</p></li>
<li><p>Team up with someone and perform hierarchical clustering on this data</p></li>
<li><p>You have been given some starter code to help you load the data</p></li>
<li><p>The data has been downloaded and processed for you (after you run the code below).</p></li>
<li><p>The data is in the variable named <code>X</code></p></li>
</ul>
<!-- TODO: find out what are labels -->
<!-- TODO: silhouette plots -->
<div id="22e4b308" class="cell" data-execution_count="1">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> requests</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.cluster <span class="im">import</span> AgglomerativeClustering</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy.cluster.hierarchy <span class="im">import</span> dendrogram, linkage</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Load data</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>X <span class="op">=</span> pd.read_csv(<span class="st">"https://raw.githubusercontent.com/cambiotraining/ml-unsupervised/refs/heads/main/course_files/data/cancer_data_saved_NC160.csv"</span>, index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Fetching labels from GitHub..."</span>)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>labs_url <span class="op">=</span> <span class="st">'https://raw.githubusercontent.com/neelsoumya/python_machine_learning/main/data/NCI60labs.csv'</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>response <span class="op">=</span> requests.get(labs_url)</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>response.raise_for_status()</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Read the raw text and split into lines.</span></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>all_lines <span class="op">=</span> response.text.strip().splitlines()</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Skip the first line (the header) to match the data dimensions.</span></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>labs <span class="op">=</span> all_lines[<span class="dv">1</span>:]</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a><span class="co"># The labels in the file are quoted (e.g., "CNS"), so we remove the quotes.</span></span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>labs <span class="op">=</span> [label.strip(<span class="st">'"'</span>) <span class="cf">for</span> label <span class="kw">in</span> labs]</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Your code below ......</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Fetching labels from GitHub...</code></pre>
</div>
</div>
<ul>
<li>Write your code while working in pairs or a group</li>
</ul>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center collapsed" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Note</span>Click to expand
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div id="f8535e5e" class="cell" data-execution_count="2">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Hierarchical Clustering</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>agg <span class="op">=</span> AgglomerativeClustering(linkage<span class="op">=</span><span class="st">'average'</span>, metric<span class="op">=</span><span class="st">'manhattan'</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>cluster_labels <span class="op">=</span> agg.fit_predict(X)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute linkage matrix for the dendrogram</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>Z <span class="op">=</span> linkage(X, method<span class="op">=</span><span class="st">'average'</span>, metric<span class="op">=</span><span class="st">'cityblock'</span>)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot Dendrogram</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>plt.figure()</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>dendrogram(Z, labels<span class="op">=</span>labs)</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">'Hierarchical Clustering Dendrogram (NCI60, Average Linkage, Manhattan Distance)'</span>)</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">'Cell Line'</span>)</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">'Distance'</span>)</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>plt.xticks(rotation<span class="op">=</span><span class="dv">90</span>)</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="evaluation_files/figure-html/cell-3-output-1.png" width="705" height="470" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<!-- end callout -->
<section id="exercise-try-another-linkage-method-and-another-distance-metric" class="level3" data-number="12.3.1">
<h3 data-number="12.3.1" class="anchored" data-anchor-id="exercise-try-another-linkage-method-and-another-distance-metric"><span class="header-section-number">12.3.1</span> Exercise: try another linkage method and another distance metric</h3>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Tip</span>Important Concept (recall)
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li><p>There is no “correct” answer in unsupervised machine learning!</p></li>
<li><p>So how do you know when you are done?</p></li>
</ul>
</div>
</div>
<!-- end callout -->
</section>
</section>
<section id="evaluating-the-quality-of-clusters" class="level2" data-number="12.4">
<h2 data-number="12.4" class="anchored" data-anchor-id="evaluating-the-quality-of-clusters"><span class="header-section-number">12.4</span> Evaluating the Quality of Clusters</h2>
<p>Evaluating the quality of clusters is a crucial step in any unsupervised learning task. Since we do not have a single <em>correct</em> answer, we use several methods that fall into three main categories:</p>
<section id="internal-evaluation" class="level3" data-number="12.4.1">
<h3 data-number="12.4.1" class="anchored" data-anchor-id="internal-evaluation"><span class="header-section-number">12.4.1</span> 1. Internal Evaluation</h3>
<p>Measures how good the clustering is based only on the data itself (e.g., how dense and well-separated the clusters are).</p>
</section>
<section id="external-evaluation" class="level3" data-number="12.4.2">
<h3 data-number="12.4.2" class="anchored" data-anchor-id="external-evaluation"><span class="header-section-number">12.4.2</span> 2. External Evaluation</h3>
<p>Measures how well the clustering results align with known, ground-truth labels. This is possible here because the NCI60 dataset has known cancer cell line types, which we loaded as <code>labs</code>.</p>
</section>
<section id="visual-evaluation" class="level3" data-number="12.4.3">
<h3 data-number="12.4.3" class="anchored" data-anchor-id="visual-evaluation"><span class="header-section-number">12.4.3</span> 3. Visual Evaluation</h3>
<p>Inspecting plots (like the dendrogram or PCA) to see if the groupings seem logical.</p>
<hr>
<p>Let us add the two most common metrics: one internal and one external.</p>
<hr>
</section>
<section id="internal-evaluation-silhouette-score" class="level3" data-number="12.4.4">
<h3 data-number="12.4.4" class="anchored" data-anchor-id="internal-evaluation-silhouette-score"><span class="header-section-number">12.4.4</span> Internal Evaluation: Silhouette Score</h3>
<p>The <strong>Silhouette Score</strong> measures how similar an object is to its own cluster (cohesion) compared to other clusters (separation).</p>
<ul>
<li><strong>Score Range</strong>: -1 to +1<br>
</li>
<li><strong>Interpretation</strong>:
<ul>
<li><strong>+1</strong>: The sample is far away from the neighboring clusters (very good).</li>
<li><strong>0</strong>: The sample is on or very close to the decision boundary between two neighboring clusters.</li>
<li><strong>-1</strong>: The sample is assigned to the wrong cluster.</li>
</ul></li>
</ul>
<hr>
</section>
<section id="external-evaluation-adjusted-rand-index-ari" class="level3" data-number="12.4.5">
<h3 data-number="12.4.5" class="anchored" data-anchor-id="external-evaluation-adjusted-rand-index-ari"><span class="header-section-number">12.4.5</span> External Evaluation: Adjusted Rand Index (ARI)</h3>
<p>The <strong>Adjusted Rand Index (ARI)</strong> measures the similarity between the true labels (<code>labs</code>) and the labels assigned by our clustering algorithm (<code>cluster_labels</code>). It accounts for chance groupings.</p>
<ul>
<li><strong>Score Range</strong>: -1 to +1<br>
</li>
<li><strong>Interpretation</strong>:
<ul>
<li><strong>+1</strong>: Perfect agreement between true and predicted labels.</li>
<li><strong>0</strong>: Random labeling (no correlation).</li>
<li><strong>&lt; 0</strong>: Worse than random labeling.</li>
</ul></li>
<li>Here is how you would implement this</li>
</ul>
<div id="bb42ac50" class="cell" data-execution_count="3">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.metrics <span class="im">import</span> silhouette_samples, silhouette_score, adjusted_rand_score <span class="co"># Import evaluation metrics</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Hierarchical Clustering</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>agg <span class="op">=</span> AgglomerativeClustering(linkage<span class="op">=</span><span class="st">'average'</span>, metric<span class="op">=</span><span class="st">'manhattan'</span>)</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>cluster_labels <span class="op">=</span> agg.fit_predict(X)</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="co"># 1. Internal Evaluation: Silhouette Score</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Measures how well-separated clusters are based on the data itself.</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>silhouette <span class="op">=</span> silhouette_score(X, cluster_labels, metric<span class="op">=</span><span class="st">'manhattan'</span>)</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Silhouette score"</span>)</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(silhouette)</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Score is from -1 to 1. Higher is better"</span>)</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a><span class="co"># 2. External Evaluation: Adjusted Rand Index</span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Compares our cluster labels to the true cancer type labels.</span></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>ari <span class="op">=</span> adjusted_rand_score(labs, cluster_labels)</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Adjusted Rand Index"</span>)</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(ari)</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Compares to true labels. Score is from -1 to 1. Higher is better"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Silhouette score
0.1436950300449066
Score is from -1 to 1. Higher is better
Adjusted Rand Index
0.0554671516253694
Compares to true labels. Score is from -1 to 1. Higher is better</code></pre>
</div>
</div>
</section>
<section id="intuition-of-silhouette-score" class="level3" data-number="12.4.6">
<h3 data-number="12.4.6" class="anchored" data-anchor-id="intuition-of-silhouette-score"><span class="header-section-number">12.4.6</span> Intuition of silhouette score</h3>
<p>Think of each data point as asking two questions:</p>
<ul>
<li><strong>How close am I to points in my own cluster?</strong> (call this <strong>a</strong>)<br>
</li>
<li><strong>How close would I be, on average, to the <em>nearest other</em> cluster?</strong> (call this <strong>b</strong>)</li>
</ul>
<p>The silhouette value for the point compares those two answers:</p>
<ul>
<li>If <strong>a</strong> is much smaller than <strong>b</strong> (you are much closer to your own cluster than to any other), the silhouette is close to <strong>+1</strong> → <em>great fit</em>.<br>
</li>
<li>If <strong>a ≈ b</strong>, silhouette is near <strong>0</strong> → <em>on the boundary between clusters</em>.<br>
</li>
<li>If <strong>a &gt; b</strong>, silhouette is <strong>negative</strong> → <em>probably misassigned</em> (you are closer to another cluster than your own).</li>
</ul>
<p>Numerically:</p>
<p>[ s = ]</p>
<p>So</p>
<p>[ s ]</p>
<!--
* Silhouette plots

A silhouette plot is a visual diagnostic for clustering quality that (1) computes a silhouette value for each sample and (2) shows the distribution of those values for every cluster. It helps you see which clusters are tight and well-separated and which contain ambiguous or poorly assigned samples.

* Notes on interpreting silhouette plots

- Each horizontal block is a cluster; the width at a given vertical position is the silhouette value of a sample.

- Values close to +1 → sample is well matched to its own cluster and poorly matched to neighbors.

- Values near 0 → sample lies between clusters.

- Negative values → sample is likely assigned to the wrong cluster.

- The red dashed line is the average silhouette score; use it as a quick summary, but always inspect per-cluster distributions — a high average can hide poorly-formed small clusters.
-->
<!--
* Notes on interpreting silhouette plots

- Each horizontal block is a cluster; the width at a given vertical position is the silhouette value of a sample.

- Values close to +1 → sample is well matched to its own cluster and poorly matched to neighbors.

- Values near 0 → sample lies between clusters.

- Negative values → sample is likely assigned to the wrong cluster.

- The red dashed line is the average silhouette score; use it as a quick summary, but always inspect per-cluster distributions — a high average can hide poorly-formed small clusters.
-->
<ul>
<li><p>Compare to literature and what others have done</p></li>
<li><p>Plain old visual evaluation</p></li>
<li><p>compare to labels of what these cell lines are (assuming this is available)</p></li>
</ul>
<!--TODO: XX what do these labels mean? -->
<!--
The silhouette value (per sample)

For sample 
𝑖
i:

𝑎
(
𝑖
)
a(i) = average distance from 
𝑖
i to all other points in the same cluster (intra-cluster distance).

For every other cluster 
𝐶
C, compute the average distance from 
𝑖
i to points in 
𝐶
C; let 
𝑏
(
𝑖
)
b(i) be the smallest of those (the nearest other cluster's average distance).

The silhouette value is

𝑠
(
𝑖
)
=
𝑏
(
𝑖
)
−
𝑎
(
𝑖
)
max
⁡
(
𝑎
(
𝑖
)
,
 
𝑏
(
𝑖
)
)
.
s(i)=
max(a(i),b(i))
b(i)−a(i)
    ​

.

Range: 
−
1
≤
𝑠
(
𝑖
)
≤
1
−1≤s(i)≤1.

𝑠
(
𝑖
)
≈
1
s(i)≈1: sample is well matched to its own cluster and far from neighbours.

𝑠
(
𝑖
)
≈
0
s(i)≈0: sample lies between two clusters.

𝑠
(
𝑖
)
<
0
s(i)<0: sample is probably assigned to the wrong cluster (closer on average to another cluster than to its own).

Example: if 
𝑎
=
0.3
a=0.3 and 
𝑏
=
0.6
b=0.6, then 
𝑠
=
(
0.6
−
0.3
)
/
0.6
=
0.5
s=(0.6−0.3)/0.6=0.5 — a reasonably good assignment.

The silhouette plot

Each cluster is shown as a horizontal block.

Within each block, samples are ordered by silhouette value and displayed as horizontal bars (width = 
𝑠
(
𝑖
)
s(i)).

A dashed vertical line shows the average silhouette score across all samples.

The plot highlights:

Cluster compactness (long bars near +1),

Presence of many low/negative values (bad clusters or misassignments),

Relative sizes of clusters.
-->
<p>How to interpret (practical heuristics)</p>
<ul>
<li><p>Mean silhouette ≳ 0.5 → strong structure (good clustering).</p></li>
<li><p>Mean silhouette ≈ 0.25–0.5 → weak to moderate structure; inspect clusters individually.</p></li>
<li><p>Mean silhouette ≲ 0.25 → little structure; clustering may be unreliable. (These are rules of thumb — context and domain knowledge matter.)</p></li>
</ul>
<!--
* Silhouette plots for hierarchical clustering

-->
<ul>
<li><p>Also compare to clusterings of other cancer cell lines</p></li>
<li><p>Does the cell line also show up in other datasets? (<code>external validation</code>)</p></li>
</ul>
<!--

# Practical code snippets (copy into a Jupyter notebook)

Replace `X` with your feature matrix (cells × genes or PCA components), and `y_true` with any ground-truth labels you might have.

```python
# Required packages
import numpy as np
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans
from sklearn.metrics import (
    silhouette_score,
    calinski_harabasz_score,
    davies_bouldin_score,
    adjusted_rand_score,
    normalized_mutual_info_score
)
from sklearn.manifold import trustworthiness
from sklearn.model_selection import train_test_split

# 1) PCA explained variance (dimensionality reduction check)
pca = PCA(n_components=20).fit(X)
print("Explained variance ratio (first 10):", pca.explained_variance_ratio_[:10])
print("Cumulative:", np.cumsum(pca.explained_variance_ratio_)[:10])

# 2) Run simple clustering and internal metrics
k = 6
km = KMeans(n_clusters=k, random_state=0).fit(X)
labels = km.labels_
print("Silhouette:", silhouette_score(X, labels))
print("Calinski-Harabasz:", calinski_harabasz_score(X, labels))
print("Davies-Bouldin:", davies_bouldin_score(X, labels))

# 3) If you have ground-truth labels (y_true)
# print("ARI:", adjusted_rand_score(y_true, labels))
# print("NMI:", normalized_mutual_info_score(y_true, labels))

# 4) Trustworthiness for an embedding (e.g., embedding = UMAP_result)
# trust = trustworthiness(X, embedding, n_neighbors=15)
# print("Trustworthiness of embedding:", trust)

# 5) Stability via subsampling
def stability_score(clustering_fn, X, n_runs=10, sample_frac=0.8):
    from sklearn.metrics import adjusted_rand_score
    labels_list = []
    rng = np.random.RandomState(0)
    for i in range(n_runs):
        idx = rng.choice(np.arange(X.shape[0]), size=int(sample_frac*X.shape[0]), replace=False)
        labels_sub = clustering_fn(X[idx])
        labels_list.append((idx, labels_sub))
    # compute mean ARI between overlapping samples
    aris = []
    for i in range(len(labels_list)):
        for j in range(i+1, len(labels_list)):
            idx_i, lab_i = labels_list[i]
            idx_j, lab_j = labels_list[j]
            common = np.intersect1d(idx_i, idx_j)
            if len(common) < 10:
                continue
            map_i = {v: k for k, v in enumerate(idx_i)}
            map_j = {v: k for k, v in enumerate(idx_j)}
            lab_i_common = [lab_i[map_i[c]] for c in common]
            lab_j_common = [lab_j[map_j[c]] for c in common]
            aris.append(adjusted_rand_score(lab_i_common, lab_j_common))
    return np.mean(aris)

# Example clustering function for kmeans on raw X
clustering_fn = lambda Xsub: KMeans(n_clusters=k, random_state=0).fit(Xsub).labels_
# print("Stability (mean ARI):", stability_score(clustering_fn, X, n_runs=6))
```

---

# Visual diagnostics 

* **Silhouette plot**: for each cluster show silhouette widths; points near 1 are clean, negative means bad assignment.
* **Elbow plot**: inertia/within-cluster sum of squares vs k (use with caution).
* **UMAP/t-SNE** colored by cluster and colored by known biological metadata (donor, batch, cell cycle) — if clusters align with batch, that's a red flag.
* **Heatmap** of marker genes (or top variable genes) with clusters as columns — helps biological interpretation.
* **Violin/boxplots** of expression of canonical marker genes across clusters.
* **Confusion matrix** comparing cluster labels vs known labels (if available).
* **Stability heatmap** of ARI between parameter settings (k values, min\_samples for DBSCAN, perplexity for t-SNE).

---

# Biological validation techniques

* Check expression of known **marker genes** per cluster — violin plots, fraction of cells expressing gene, median expression.
* **Enrichment tests**: test if clusters are enriched for GO terms, pathways, or cell type marker sets.
* Check **batch/donor composition** per cluster — if one donor dominates a cluster, suspect batch effect.
* Use **pairwise marker identification** (differential expression between clusters) and then ask: does this match known biology?

---

# Hands-on classroom plan (90–120 minutes)

1. **10 min** — Intuition + categories (internal / external / biological).
2. **20 min** — Demo: run k-means on a tiny single-cell toy (PCA → kmeans), compute silhouette / CH / DB. Show silhouette plot and UMAP colored by clusters.
3. **25 min** — Students exercise (paired): try k = 2..10, make elbow and silhouette; pick “best” k and justify biologically.
4. **20 min** — Introduce external metrics (ARI/NMI). Give a partial ground truth (e.g., FACS labels) and compare clusterings. Interpret mismatch.
5. **15 min** — Stability exercise: subsample and compute ARI between runs. Discuss parameter sensitivity (UMAP perplexity/embedding variance).
6. **Final 10 min** — Discuss pitfalls & how to report results.

---

# Example mini-assignments (graded)

1. **Compare 3 clustering algorithms** (k-means, hierarchical, DBSCAN) on a provided scRNA dataset: produce internal metric table, UMAPs, marker heatmap, and short writeup (max 500 words) interpreting differences and recommending one method.
2. **Stability report**: for chosen clustering, show results of 10 subsamples, plot ARI distribution, and explain what it implies for reproducibility.
3. **Biological validation**: pick 5 marker genes and show their distribution across clusters; perform enrichment test and conclude whether clusters are biologically meaningful.

---

# How to grade / rubric (suggested)

* **40%** biological interpretability (marker genes, enrichment)
* **30%** technical diagnostics (internal metrics, stability)
* **20%** visualization & clarity (UMAPs, heatmaps, silhouette plot)
* **10%** reproducibility (notebooks that run, parameter reporting)

---

# Common pitfalls to teach explicitly

* Relying on a single metric (e.g., optimizing silhouette only) — always combine metrics + biology.
* Over-interpreting t-SNE/UMAP global structure — these emphasize local neighborhoods; check trustworthiness.
* Ignoring batch/donor effects — clusters that reflect batches are misleading.
* Cherry-picking k or parameters to match prior beliefs — require pre-registered analysis choices or show sensitivity analyses.

---

# Quick “cheat-sheet” (one-paragraph version to hand out)

* **No ground truth?** Use internal metrics (silhouette, CH, DB) + stability checks + marker gene verification.
* **Have partial labels?** Use ARI/NMI to compare and treat labels as guides, not absolute truth.
* **Embedding checks:** report PCA explained variance and trustworthiness for UMAP/t-SNE.
* **Visualize:** always show embedding + gene heatmap + violin plots for markers.
* **Reproducibility:** report parameters and run at least one sensitivity/subsampling analysis.

---


-->
</section>
</section>
<section id="summary" class="level2" data-number="12.5">
<h2 data-number="12.5" class="anchored" data-anchor-id="summary"><span class="header-section-number">12.5</span> Summary</h2>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Tip</span>Key Points
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>We learnt evaluation is difficult in unsupervised machine learning!</li>
</ul>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../materials/missing_data.html" class="pagination-link" aria-label="Missing values in unsupervised machine learning">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Missing values in unsupervised machine learning</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../materials/pca_tsne_notwork.html" class="pagination-link" aria-label="When PCA or tSNE might not work">
        <span class="nav-page-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">When PCA or tSNE might not work</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>Licensed <a href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a> <img src="https://creativecommons.org/images/deed/cc_icon_black_x2.png" class="img-fluid" width="25"> Cambridge Research Informatics Training</p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>